Bot initiates training on 0 energy

This occurs about 5 times throughout this run each with the 'Don't Overdo It' event, each at 0% energy during the turn. There are also some questionable failed training events with 'Get Well Soon' in the logs as well.

Maximum Failure Chance Allowed: 15%
Risky Training Maximum Failure Chance: 25%

[Seiun Sky_2026-04-11-14-28-23.txt](https://github.com/user-attachments/files/26646764/Seiun.Sky_2026-04-11-14-28-23.txt)

```
00:21:46.553 [INFO] [FINALS] Checking if bot should stop before the finals.
00:21:46.555 [VERBOSE] [TRAINEE] Name: Seiun Sky
00:21:46.555 [VERBOSE] [TRAINEE] Stats: Spd=336, Sta=311, Pow=384, Gut=215, Wit=375
00:21:46.555 [VERBOSE] [TRAINEE] Energy: 0%
00:21:46.555 [VERBOSE] [TRAINEE] Fans: 74315
00:21:46.557 [VERBOSE] [TRAINEE] Skill Points: 71
00:21:46.557 [VERBOSE] [TRAINEE] Track: Turf=A, Dirt=G
00:21:46.557 [VERBOSE] [TRAINEE] Distance: Sprint=G, Mile=A, Medium=A, Long=A
00:21:46.558 [VERBOSE] [TRAINEE] Style: Front=A, Pace=B, Late=D, End=E
00:21:46.558 [INFO] [TRACKBLAZER] Skipping Training Items dialog as no relevant items are in the cached inventory.
00:21:46.766 [INFO] [TRACKBLAZER] Evaluating for Irregular Training...
00:21:47.335 [INFO] [LOADING] Now checking if the game is still loading...

00:21:47.556 [VERBOSE] [TRAINING] Now starting process to analyze all 5 Trainings.
00:21:48.178 [INFO] [LOADING] Now checking if the game is still loading...
00:21:49.467 [INFO] [TRAINING] Flag set to ignore failure chance. Proceeding to acquire all other percentages and total stat increases...
00:21:49.877 [INFO] [LOADING] Now checking if the game is still loading...
00:21:53.906 [WARN] [WARN] goToStat:: Failed to go to GUTS on training screen after 3 attempts.
00:21:53.906 [ERROR] [ERROR] analyzeTrainings:: Failed to click training button for GUTS. Aborting training...

00:21:53.907 [VERBOSE] ========== Training Analysis Results ==========
Scoring Mode: Trackblazer (Irregular Training)
Current Date: CLASSIC YEAR EARLY JANUARY (Turn 25)
Current Stats: Speed=336, Stam=311, Pow=384, Guts=215, Wit=375
Stat Targets (MEDIUM): Speed=1100, Stam=700, Pow=900, Guts=500, Wit=800
Completion: SPEED=31%, STAMINA=44%, POWER=43%, GUTS=43%, WIT=47%

Energy recovery needed. No analysis performed.
================================================

00:21:53.908 [INFO] [TRACKBLAZER] No valid Irregular Training found. Backing out to resume racing logic.
00:21:54.423 [INFO] [LOADING] Now checking if the game is still loading...

00:21:54.757 [INFO] [INJURY] Checking if there is an injury that needs healing on CLASSIC YEAR EARLY JANUARY (Turn 25).
00:21:54.822 [WARN] [WARN] checkInjury:: Failed to detect the Infirmary button.

00:21:54.823 [INFO] [RACE] Now determining eligibility to start the extra racing process...
00:21:55.275 [INFO] [INFO] Detected day for extra racing: 24
00:21:55.275 [INFO] [RACE] Current remaining number of days before the next mandatory race: 24.
00:21:55.275 [INFO] [RACE] Trackblazer scenario detected. Bypassing smart racing and interval checks.
00:21:55.370 [INFO] [INFO] Bot has no injuries, mood is sufficient and extra races can be run today. Setting the action to RACE.
00:21:55.468 [INFO] [INFO] All checks are cleared for racing.
00:21:55.634 [INFO] [TRACKBLAZER] Checking for suitable races.
00:21:55.804 [ERROR] [ERROR] handleRaceEvents:: Failed to click Races button.
00:21:57.110 [INFO] [LOADING] Now checking if the game is still loading...
00:21:57.289 [INFO] [TRACKBLAZER] Starting specialized Training process.
00:21:57.368 [ERROR] [ERROR] handleTrackblazerTraining:: Failed to enter Training screen.

00:21:58.053 [INFO] [INFO] Checking if the bot is sitting on the Training Event screen.
00:21:58.135 [VERBOSE] [INFO] Bot is at the Training Event screen.

00:21:58.136 [VERBOSE] ********************
00:21:58.136 [VERBOSE] [TRAINING_EVENT] Starting Training Event process on CLASSIC YEAR EARLY JANUARY (Turn 25).
00:21:58.784 [INFO] [INFO] Detected event title text with Tesseract: Don't Overdo It!
00:21:58.785 [INFO] [TRAINING_EVENT_RECOGNIZER] Now starting process to find most similar string to: Don't Overdo It!
00:21:58.786 [INFO] [TRAINING_EVENT_RECOGNIZER] Detected special event pattern: Don't Overdo It!. Will restrict search to this event.
00:21:58.786 [INFO] [TRAINING_EVENT_RECOGNIZER] Finished process to find similar string.
00:21:58.786 [INFO] [TRAINING_EVENT_RECOGNIZER] Event data fetched for "Don't Overdo It!".
00:21:58.786 [INFO] [TRAINING_EVENT_RECOGNIZER] Special event "Don't Overdo It!" detected.
00:21:58.786 [VERBOSE] [TRAINING_EVENT] Detected special event: Don't Overdo It!
00:21:58.787 [VERBOSE] [TRAINING_EVENT] Using setting: Option 2: (Random) Mood -3 / Stat decrease / Get Practice Poor negative status (Option 2)
00:21:58.787 [VERBOSE] [TRAINING_EVENT] Special event override applied: option 2
00:21:58.887 [INFO] [LOADING] Now checking if the game is still loading...
00:21:59.857 [VERBOSE] [TRAINING_EVENT] Process to handle detected Training Event completed.
00:21:59.858 [VERBOSE] ********************
```

A secondary failure check using the energy bar remaining values (backup if the OCR doesn't detect the failure rates), with 2% failure added for every 1 energy below 50 energy would be a good start.

Another suggestion for training chances would be to implement a dynamic one based on the selected training main stat, for example +20 to speed,  accepts up to 10% failure, +50 up to 25%, +100 to 50% etc. (atp I would just use energy or charm past 25%)

--- a/android/app/src/main/java/com/steve1316/uma_android_automation/bot/Training.kt
+++ b/android/app/src/main/java/com/steve1316/uma_android_automation/bot/Training.kt
@@ -830,6 +830,15 @@ class Training(private val game: Game, private val campaign: Campaign) {
         val ignoreFailureChance = args["ignoreFailureChance"] as? Boolean ?: false
         val isIrregularEvaluation = args["isIrregularEvaluation"] as? Boolean ?: false
 
+        // Skip training analysis entirely when energy is depleted and no charm is available to offset the failure chance.
+        if (!test && !ignoreFailureChance && !campaign.checkFinals() && campaign.trainee.energy <= 0) {
+            MessageLog.i(TAG, "[TRAINING] Skipping training analysis as energy is ${campaign.trainee.energy}% with no Good-Luck Charm to offset failure chance.")
+            needsEnergyRecovery = true
+            trainingMap.clear()
+            skippedTrainingMap.clear()
+            return
+        }
+
         if (test) {
             MessageLog.v(TAG, "\n[TRAINING] Now starting process to analyze all 5 Trainings for Testing.")
         } else if (singleTraining) {

--- a/android/app/src/main/java/com/steve1316/uma_android_automation/bot/campaigns/Trackblazer.kt
+++ b/android/app/src/main/java/com/steve1316/uma_android_automation/bot/campaigns/Trackblazer.kt
@@ -686,9 +686,11 @@ class Trackblazer(game: Game) : Campaign(game) {
             val isMandatoryRace = IconRaceDayRibbon.check(game.imageUtils) || IconGoalRibbon.check(game.imageUtils)
 
             if (!isScheduledRace && !isMandatoryRace) {
-                MessageLog.i(TAG, "[TRACKBLAZER] Evaluating for Irregular Training...")
-
-                if (ButtonTraining.click(game.imageUtils)) {
+                // Skip irregular training evaluation when energy is depleted and no charm can offset the failure chance.
+                if (trainee.energy <= 0 && !hasCharmAvailable) {
+                    MessageLog.i(TAG, "[TRACKBLAZER] Skipping Irregular Training evaluation as energy is ${trainee.energy}% with no Good-Luck Charm available.")
+                    bHasCheckedIrregularTrainingThisTurn = true
+                } else if (ButtonTraining.click(game.imageUtils)) {
                     game.wait(game.dialogWaitDelay)
 
                     val isIrregularEvaluation = true

